src/Java.Interop.BootstrapTasks/Java.Interop.BootstrapTasks/JdkDirectoryFinder.cs

```diff
@@ -0,0 +1,62 @@
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Text.RegularExpressions;
+
+using Cadenza;
+
+using Microsoft.Build.Utilities;
+using Microsoft.Build.Framework;
+
+namespace Java.Interop.BootstrapTasks {
+
+	public class JdkDirectoryFinder : Task {
+
+		[Required]
+		public  string  JdksRoot              { get; set; }
+
+		public  string  MaximumJdkVersion     { get; set; }
+
+		[Output]
+		public  string  PreferredJdkRoot      { get; set; }
+
+		static  Regex   VersionExtractor  = new Regex (@"(?<version>[\d]+(\.\d+)+)", RegexOptions.Compiled);
+
+		public override bool Execute ()
+		{
+			var version = GetMaxJdkVersion ();
+			var dirs    = new List<string>();
+			foreach (var d in Directory.EnumerateDirectories (JdksRoot)) {
+				var n = Path.GetFileName (d);
+				var m = version == null ? null : VersionExtractor.Match (n);
+				Version v;
+
+				if (version == null || !m.Success || !Version.TryParse (m.Groups ["version"].Value, out v)) {
+					dirs.Add (d);
+					continue;
+				}
+				if (v <= version) {
+					dirs.Add (d);
+				}
+			}
+			dirs.Sort (NaturalStringComparer.Default);
```

Comment:
Could we instead parse out to `System.Version` and sort by that? Then we wouldn't need the `NaturalStringComparer`.

LINQ would be ok here, to do something like:
```csharp
var dirs = from d in Directory.EnumerateDirectories (JdksRoot)
                let n = Path.GetFileName (d)
                let m = VersionExtracter.Match (n)
                let r = Version.TryParse (m.Groups ["version"].Value, out Version v)
                where r && (maxVersion == null || v < maxVersion)
                order by v desc
                select d;
PreferredJdkRoot = dirs.FirstOrDefault ();
```